src/check/check_types/occurs.zig

```diff
@@ -0,0 +1,785 @@
+            return err;
+        };
+    }
+
+    /// Check if a slice of sub vars are recursive
+    /// In the event of an error, append the root var to the chain
+    fn occursSubVars(self: *Self, root_var: ResolvedVarDesc, sub_vars: []Var, ctx: Context) Error!void {
+        for (sub_vars) |sub_var| {
+            try self.occursSubVar(root_var, sub_var, ctx);
+        }
+    }
+};
+
+/// This type represents the context of a recursive branch in the occurs check
+///
+/// This types is passed down through the occurs check and is modified to keep
+/// track of whatever we've seen so far in this branch.
+const Context = struct {
+    recursion_allowed: bool,
+    seen_nominal: bool,
+
+    fn init() Context {
+        return .{ .recursion_allowed = false, .seen_nominal = false };
+    }
+
+    /// Mark that recursion is allowed
+    /// ie the chain passes through a Box, List or Tag Union
+    fn allowRecursion(self: Context) Context {
+        return .{
+            .recursion_allowed = true,
+            .seen_nominal = self.seen_nominal,
+        };
+    }
+
+    /// Mark that we have seen a nominal type
+    fn markNominal(self: Context) Context {
+        return .{
+            .recursion_allowed = self.recursion_allowed,
+            .seen_nominal = true,
+        };
+    }
+};
+
+/// Struct to hold intermediate values used during occurs check
+const Scratch = struct {
+    const Self = @This();
+
+    gpa: std.mem.Allocator,
+
+    seen: Var.SafeList,
+    err_chain: Var.SafeList,
+    err_chain_nominal_vars: Var.SafeList,
+    visited: MkSafeList(DescStoreIdx),
+
+    fn init(gpa: std.mem.Allocator) Self {
+        // TODO: eventually use herusitics here to determine sensible defaults
+        // Rust compiler inits with 1024 capacity. But that feels like a lot.
+        // Typical recursion cases should only be a few layers deep?
+        return .{
+            .gpa = gpa,
+            .seen = Var.SafeList.initCapacity(gpa, 32),
+            .err_chain = Var.SafeList.initCapacity(gpa, 32),
+            .err_chain_nominal_vars = Var.SafeList.initCapacity(gpa, 8),
+            .visited = MkSafeList(DescStoreIdx).initCapacity(gpa, 64),
+        };
+    }
+
+    fn deinit(self: *Self) void {
+        self.seen.deinit(self.gpa);
+        self.err_chain.deinit(self.gpa);
+        self.err_chain_nominal_vars.deinit(self.gpa);
+        self.visited.deinit(self.gpa);
+    }
+
+    fn reset(self: *Self) void {
+        self.seen.items.clearRetainingCapacity();
+        self.err_chain.items.clearRetainingCapacity();
+        self.err_chain_nominal_vars.items.clearRetainingCapacity();
+        self.visited.items.clearRetainingCapacity();
+    }
+
+    fn hasSeenVar(self: *const Self, var_: Var) bool {
+        for (self.seen.items.items) |seen_var| {
+            if (seen_var == var_) return true;
+        }
+        return false;
+    }
+
+    fn appendSeen(self: *Self, var_: Var) void {
+        _ = self.seen.append(self.gpa, var_);
```

Comment:
In working on `unify` I've actually personally been liking having things do `std.mem.Allocation.Error!void` and actually use `try` (instead of the exit on oom thing) - what do you think of doing more of that?

Reply:
I'm into that, imo the more explicit the points of failure are the better. And tracking the error in the type sigs would make it easier to understand what's allocating and what's not, at a glance